mythril/support/signatures.py

```diff
@@ -24,46 +24,34 @@
     FourByteDirectoryOnlineLookupError = Exception
 
 
-class SimpleFileLock(object):
-    # todo: replace with something more reliable. this is a quick shot on concurrency and might not work in all cases
-
-    def __init__(self, path):
-        self.path = path
-        self.lockfile = pathlib.Path("%s.lck" % path)
-        self.locked = False
-
-    def aquire(self, timeout=5):
-        if self.locked:
-            raise Exception("SimpleFileLock: lock already aquired")
+try:
+    # Posix based file locking (Linux, Ubuntu, MacOS, etc.)
+    import fcntl
 
-        t_end = time.time() + timeout
-        while time.time() < t_end:
-            # try to aquire lock
-            try:
-                self.lockfile.touch(mode=0o0000, exist_ok=False)  # touch the lockfile
-                # lockfile does not exist. we have a lock now
-                self.locked = True
-                return
-            except FileExistsError as fee:
-                # check if lockfile date exceeds age and cleanup lock
-                if time.time() > self.lockfile.stat().st_mtime + 60 * 5:
-                    self.release(force=True)  # cleanup old lockfile > 5mins
+    def lock_file(f, exclusive=False):
+        if f.mode == 'r' and exclusive:
+            raise Exception('Please use non exclusive mode for reading')
+        flag = fcntl.LOCK_EX if exclusive else fcntl.LOCK_SH
+        fcntl.lockf(f, flag)
 
-                time.sleep(0.5)  # busywait is evil
-                continue
+    def unlock_file(f):
+        return
 
-        raise Exception("SimpleFileLock: timeout hit. failed to aquire lock: %s" % (time.time() - self.lockfile.stat().st_mtime))
+except ImportError:
+    # Windows file locking
+    # TODO: confirm the existence or non existence of shared locks in windows msvcrt and make changes based on that
+    import msvcrt
 
-    def release(self, force=False):
-        if not force and not self.locked:
-            raise Exception("SimpleFileLock: aquire lock first")
+    def file_size(f):
+        return os.path.getsize(os.path.realpath(f.name))
 
-        try:
-            self.lockfile.unlink()  # might throw if we force unlock and the file gets removed in the meantime. TOCTOU
-        except FileNotFoundError as fnfe:
-            logging.warning("SimpleFileLock: release(force=%s) on unavailable file. race? %r" % (force, fnfe))
+    def lock_file(f, exclusive=False):
+        if f.mode == 'r' and exclusive:
+            raise Exception('Please use non exclusive mode for reading')
+        msvcrt.locking(f.fileno(), msvcrt.LK_RLCK, file_size(f))
```

Comment:
There seems to be no difference between exclusive and non exclusive locks, that shouldn't be right right?

Reply:
Thats for windows, i didn't find any flag for shared locks, so i put it as a TODO.

Reply:
I'm not 100% sure but I think by default in windows you have an "exclusive" lock, and you can choose to make it shared. (Don't know how the lib implements that though :sweat_smile: )

Reply:
The library doesn't seem to have any shared lock.Something more should be done for it. and also there doesn't seem to be any performance difference with either of the locks, so i thought it might be better to leave it for now.